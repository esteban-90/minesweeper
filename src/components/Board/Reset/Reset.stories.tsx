import type { Story, Meta } from '@storybook/react'
import { Reset, type ResetProps } from './Reset'

export default {
  title: 'Components/Board/Reset',
  component: Reset,
} as Meta

const Template: Story<ResetProps> = (args) => <Reset {...args} />

export const ResetExample = Template.bind({})
