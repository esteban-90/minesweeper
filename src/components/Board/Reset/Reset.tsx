import type { FC } from 'react'
import styled from '@emotion/styled'
import { usePressedButton } from '@/hooks'

export interface ResetProps {
  /**
   * Reset action handler
   */
  reset: () => void
}

export const Reset: FC<ResetProps> = ({ reset }): JSX.Element => {
  const [pressed, down, up] = usePressedButton()

  return (
    <Button onClick={reset} onMouseDown={down} onMouseUp={up} onMouseLeave={up}>
      {pressed ? 'ðŸ˜¯' : 'ðŸ™‚'}
    </Button>
  )
}

const Button = styled.button`
  font-size: 1.5vw;
  cursor: pointer;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  background-color: #d1d1d1;
  border-color: white #9e9e9e #9e9e9e white;
  padding: 0;
`
