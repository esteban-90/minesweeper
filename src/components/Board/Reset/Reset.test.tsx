import { render, screen, fireEvent } from '@testing-library/react'
import { Reset } from './Reset'

describe('Reset test cases:', () => {
  it('reset handler should be called', () => {
    const reset = vitest.fn()
    render(<Reset reset={reset} />)

    const smilingFace = screen.getByText('ðŸ™‚')
    fireEvent.click(smilingFace)

    expect(reset).toBeCalled()
  })

  describe('without reset handler:', () => {
    beforeEach(() => {
      render(<Reset reset={() => null} />)
    })

    it('should render elements with default state', () => {
      const smilingFace = screen.getByText('ðŸ™‚')
      expect(smilingFace).toBeInTheDocument()
    })

    it('should change state when onMouseDown and onMouseUp events happen', () => {
      const smilingFace = screen.getByText('ðŸ™‚')
      fireEvent.mouseDown(smilingFace)

      const hushedFace = screen.getByText('ðŸ˜¯')
      expect(hushedFace).toBeInTheDocument()

      fireEvent.mouseUp(hushedFace)
      expect(smilingFace).toBeInTheDocument()
    })

    it('should change state when onMouseDown and onMouseLeave events happen', () => {
      const smilingFace = screen.getByText('ðŸ™‚')
      fireEvent.mouseDown(smilingFace)

      const hushedFace = screen.getByText('ðŸ˜¯')
      expect(hushedFace).toBeInTheDocument()

      fireEvent.mouseLeave(hushedFace)
      expect(smilingFace).toBeInTheDocument()
    })
  })
})
