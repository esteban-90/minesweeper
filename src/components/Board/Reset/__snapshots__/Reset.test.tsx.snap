// Vitest Snapshot v1

exports[`Reset test cases: > should render correctly 1`] = `
<DocumentFragment>
  <button
    class="css-mdm7pv"
  >
    ğŸ™‚
  </button>
</DocumentFragment>
`;
